core/tests/LowLevelTests/Op.hs

```diff
@@ -1,60 +1,81 @@
+import Control.Exception
+import Control.Monad
+import Data.ByteString (ByteString)
+import Network.GRPC.LowLevel
+import Network.GRPC.LowLevel.Call
+import Network.GRPC.LowLevel.Client
+import Network.GRPC.LowLevel.Op
+import Network.GRPC.LowLevel.Server
+import Test.Tasty
+import Test.Tasty.HUnit as HU (testCase, (@?=))
 
 lowLevelOpTests :: TestTree
-lowLevelOpTests = testGroup "Synchronous unit tests of low-level Op interface"
-  [testCancelFromServer]
+lowLevelOpTests =
+  testGroup
+    "Synchronous unit tests of low-level Op interface"
+    [testCancelFromServer]
 
 testCancelFromServer :: TestTree
 testCancelFromServer =
   testCase "Client/Server - client receives server cancellation" $
-  runSerialTest $ \grpc ->
-    withClientServerUnaryCall grpc $
-    \(Client{..}, Server{..}, ClientCall{..}, sc@ServerCall{..}) -> do
-      serverCallCancel sc StatusPermissionDenied "TestStatus"
-      clientRes <- runOps unsafeCC clientCQ clientRecvOps
-      case clientRes of
-        Left x -> error $ "Client recv error: " ++ show x
-        Right [_,_,OpRecvStatusOnClientResult _ code _details] -> do
-          code @?= StatusPermissionDenied
-          return $ Right ()
-        wrong -> error $ "Unexpected op results: " ++ show wrong
-
+    runSerialTest $ \grpc ->
+      withClientServerUnaryCall grpc $
+        \(Client {..}, Server {..}, ClientCall {..}, sc@ServerCall {..}) -> do
+          serverCallCancel sc StatusPermissionDenied "TestStatus"
+          clientRes <- runOps unsafeCC clientCQ clientRecvOps
+          case clientRes of
+            Left x -> error $ "Client recv error: " ++ show x
+            Right [_, _, OpRecvStatusOnClientResult _ code _details] -> do
+              code @?= StatusPermissionDenied
+              return $ Right ()
+            wrong -> error $ "Unexpected op results: " ++ show wrong
 
 runSerialTest :: (GRPC -> IO (Either GRPCIOError ())) -> IO ()
 runSerialTest f =
-  withGRPC f >>= \case Left x -> error $ show x
-                       Right () -> return ()
+  withGRPC f >>= \case
+    Left x -> error $ show x
+    Right () -> return ()
 
-withClientServerUnaryCall :: GRPC
-                             -> ((Client, Server, ClientCall,
-                                  ServerCall ByteString)
-                                 -> IO (Either GRPCIOError a))
-                             -> IO (Either GRPCIOError a)
+withClientServerUnaryCall ::
+  GRPC ->
+  ( ( Client,
+      Server,
+      ClientCall,
+      ServerCall ByteString
+    ) ->
+    IO (Either GRPCIOError a)
+  ) ->
+  IO (Either GRPCIOError a)
 withClientServerUnaryCall grpc f = do
   withClient grpc clientConf $ \c -> do
     crm <- clientRegisterMethodNormal c "/foo"
-    withServer grpc serverConf $ \s ->
-      withClientCall c crm 10 $ \cc -> do
+    withServer grpc serverConf $ \s -> do
+      ccVar <- newEmptyMVar
+      bracket newEmptyMVar (\v -> putMVar v ()) $ \finished -> do
+        _ <- forkIO $
+          void $
```

Comment:
Why does this logic now require forking the `runOps` call?

Reply:
The logic relied on the fact that you could send an empty message in `runOps` while the rceiving side wasn’t listening. This seems to deadlock now. I couldn’t figure out a way to avoid that so instead we now set up the receiving side concurrently.